### PR TITLE
[Toolbars] feat: new Inline Plugin Toolbar (for plugin link)

### PR DESCRIPTION
**Pull request by noamtwix**
Click below to open examples:
rich-content: https://rich-content-Mar8-linkToolbarOption4.surge.sh/
rich-content-storybook: https://rich-content-storybook-Mar8-linkToolbarOption4.surge.sh/

### CODE REVIEW

**Comment by oded-soffrin** [comment]
In progress still?

File: packages/plugin-link/web/src/toolbar/TextLinkPanel.jsx [hunk 1]
```diff
@@ -58,10 +57,11 @@ export default class TextLinkPanel extends Component {
   };
 
   deleteLink = () => {
-    const { getEditorState, setEditorState } = this.props;
+    const { getEditorState, setEditorState, commonPubsub } = this.props;
     const editorState = getEditorState();
     const newEditorState = removeLinksInSelection(editorState, setEditorState);
     setEditorState(newEditorState);
+    commonPubsub && commonPubsub.set('cursorOnInlinePlugin', null);
```
Comment (noamtwix) [inline packages/plugin-link/web/src/toolbar/TextLinkPanel.jsx:64]: check y onChange is not enough

File: packages/plugin-link/web/src/createLinkPlugin.js [hunk 2]
```diff
@@ -43,6 +45,13 @@ const createLinkPlugin = (config = {}) => {
   };
 
   const onChange = editorState => {
+    const selection = editorState.getSelection();
+    if (hasLinksInSelection(editorState) && selection.isCollapsed()) {
+      const boundingRect = getVisibleSelectionRect(window);
+      commonPubsub.set('cursorOnInlinePlugin', { type, editorState, boundingRect });
+    } else {
+      commonPubsub.set('cursorOnInlinePlugin', null);
```
Comment (noamtwix) [inline packages/plugin-link/web/src/createLinkPlugin.js:53]: remove commonPubsub.set to the bottom

File: packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
+/* eslint-disable react/no-find-dom-node */
+import React, { Component } from 'react';
+import { findDOMNode } from 'react-dom';
+import classNames from 'classnames';
+import Separator from '../Components/Separator';
+import { BUTTONS } from './buttons';
+import toolbarStyles from '../../statics/styles/plugin-toolbar.scss';
+import RenderToolbarContent, {
+  getInitialState,
+  constructor,
```
Comment (noamtwix) [inline packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx:10]: rename

File: packages/editor-common/web/src/Base/createAtomicPluginToolbar.jsx [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+            toolbarTheme && toolbarTheme.pluginToolbar
+          ),
+          'data-hook': name ? `${name}PluginToolbar` : null,
+        };
+
+        if (this.ToolbarDecoration) {
+          const { ToolbarDecoration } = this;
+          return (
+            <ToolbarDecoration {...props}>
+              <RenderToolbarContent {...renderToolbarContentProps} />
```
Comment (noamtwix) [inline packages/editor-common/web/src/Base/createAtomicPluginToolbar.jsx:398]: create Toolbars folder
insert the toolbars components
split createBaseToolbar to: 
1. RenderToolbarContent 
2. ToolbarUtils

File: packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    scrollToolbar(event, leftDirection) {
+      event.preventDefault();
+      const { clientWidth, scrollWidth } = this.scrollContainer;
+      this.scrollContainer.scrollLeft = leftDirection
+        ? 0
+        : Math.min(this.scrollContainer.scrollLeft + clientWidth, scrollWidth);
+    }
+
+    /*eslint-disable complexity*/
+    renderButton = (button, key, themedStyle, separatorClassNames) => {
```
Comment (noamtwix) [inline packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx:124]: create new Component Button
import this Component from <RenderToolbarContent />

File: packages/plugin-link/web/src/icons/RemoveLinkIcon.jsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable max-len */
```
Comment (noamtwix) [inline packages/plugin-link/web/src/icons/RemoveLinkIcon.jsx:1]: replace Component with svg

File: packages/plugin-link/web/src/toolbar/TextLinkPanel.jsx (hunk 1 above)
Comment (noamtwix) [reply packages/plugin-link/web/src/toolbar/TextLinkPanel.jsx:64]: the cursor didn't change so onChange didn't call

File: packages/plugin-link/web/src/createLinkPlugin.js [hunk 7]
```diff
@@ -62,6 +64,15 @@ const createLinkPlugin = (config = {}) => {
   };
```
Comment (noamtwix) [inline packages/plugin-link/web/src/createLinkPlugin.js:65]: try to create one function to open and one to close the toolbar and use those functions across plugin-link

File: packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        structure: this.structure,
+      };
+
+      if (!this.shouldCreate) {
+        return null;
+      }
+
+      const { toolbarStyles: toolbarTheme } = theme || {};
+
+      // TODO: visibilityFn params?
```
[the remark this answers is not in the document]
Comment (noamtwix) [reply packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx]: This comment was there and I don't know if it's ok to remove it.
What do you think? I can remove it?

[the remark this answers is not in the document]
Comment (shaulgo) [reply packages/editor-common/web/src/Base/createInlinePluginToolbar.jsx]: check git blame (go enough backward until you find who made the comment). Check with them.

File: packages/plugin-link/web/src/toolbar/UrlLinkButton.jsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import { normalizeUrl, mergeStyles } from 'wix-rich-content-common';
+
+export default class UrlLinkButton extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { styles };
+  }
+
+  componentDidMount() {
+    const theme = this.props.theme;
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/UrlLinkButton.jsx:14]: move to constructor

File: packages/plugin-link/web/src/toolbar/UrlLinkButton.jsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    return (
+      <div className={this.state.styles.toolbarUrlContainer}>
+        <a {...anchorProps}>{href}</a>
+      </div>
+    );
+  }
+}
+
+UrlLinkButton.propTypes = {
+  getEditorState: PropTypes.func.isRequired,
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/UrlLinkButton.jsx:41]: cleanup

File: packages/plugin-link/web/src/toolbar/createLinkToolbar.js [hunk 11]
```diff
@@ -43,10 +51,18 @@ const openLinkModal = ({
   }
 };
 
-export default config => ({
+export default (config, closeInlinePluginToolbar) => ({
   TextButtonMapper: () => ({
     Link: {
-      component: TextLinkButton,
+      component: props => (
+        <TextLinkButton
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/createLinkToolbar.js:58]: use decorateComponentWithProps

File: packages/plugin-link/web/src/toolbar/inline-buttons.jsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import { BUTTONS, updateLinkAtCurrentSelection, EditIcon } from 'wix-rich-content-editor-common';
+import TextLinkButton from './TextLinkButton';
+import RemoveLinkButton from './RemoveLinkButton';
+import UrlLinkButton from './UrlLinkButton';
+
+export default (config, closeInlinePluginToolbar) => {
+  return [
+    {
+      keyName: 'url',
+      component: props => <UrlLinkButton {...config} {...props} />,
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/inline-buttons.jsx:11]: check for all of this decorateComponentWithProps

File: packages/editor-common/web/src/Base/toolbars/createInlinePluginToolbar.jsx [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    cursorIsOnInlinePlugin = () => {
+      if (
+        commonPubsub.get('cursorOnInlinePlugin') &&
+        commonPubsub.get('cursorOnInlinePlugin').boundingRect &&
+        name.toUpperCase() === commonPubsub.get('cursorOnInlinePlugin').type
+      ) {
+        this.showToolbar();
+      } else if (!commonPubsub.get('cursorOnInlinePlugin')) {
```
[the remark this answers is not in the document]
Comment (noamtwix) [reply packages/editor-common/web/src/Base/toolbars/createInlinePluginToolbar.jsx:62]: this is necessary

File: packages/plugin-link/web/src/toolbar/createLinkToolbar.js (hunk 11 above)
Comment (noamtwix) [reply packages/plugin-link/web/src/toolbar/createLinkToolbar.js:58]: it's doesn't seem to work

File: packages/plugin-link/web/src/toolbar/inline-buttons.jsx (hunk 12 above)
Comment (noamtwix) [reply packages/plugin-link/web/src/toolbar/inline-buttons.jsx:11]: doesn't work

File: packages/editor-common/web/src/Base/toolbars/createInlinePluginToolbar.jsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (shaulgo) [reply packages/editor-common/web/src/Base/toolbars/createInlinePluginToolbar.jsx:59]: why?

**Pull request merged**